redash/schema.py

```diff
@@ -0,0 +1,186 @@
+        table_data[persisted_table.name]["id"] = persisted_table.id
+        persisted_table_data.append(table_data[persisted_table.name])
+
+    models.db.session.bulk_update_mappings(TableMetadata, persisted_table_data)
+
+    # Find the tables that need to be created by subtracting the sets:
+    persisted_table_set = set([col_data["name"] for col_data in persisted_table_data])
+    tables_to_create = existing_tables_set.difference(persisted_table_set)
+
+    table_metadata = [table_data[table_name] for table_name in tables_to_create]
+
+    models.db.session.bulk_insert_mappings(TableMetadata, table_metadata)
+
+
+def insert_or_update_column_metadata(table, existing_columns_set, column_data):
+    persisted_columns = ColumnMetadata.query.filter(
+        ColumnMetadata.name.in_(existing_columns_set),
+        ColumnMetadata.table_id == table.id,
+    ).all()
+
+    persisted_column_data = []
+    for persisted_column in persisted_columns:
+        # Add IDs to persisted column data so it can be used for updates.
+        column_data[persisted_column.name]["id"] = persisted_column.id
+        persisted_column_data.append(column_data[persisted_column.name])
+
+    models.db.session.bulk_update_mappings(ColumnMetadata, persisted_column_data)
+
+    # Find the columns that need to be created by subtracting the sets:
+    persisted_column_set = set([col_data["name"] for col_data in persisted_column_data])
+    columns_to_create = existing_columns_set.difference(persisted_column_set)
+
+    column_metadata = [column_data[col_name] for col_name in columns_to_create]
+
+    models.db.session.bulk_insert_mappings(ColumnMetadata, column_metadata)
+
+
+class SchemaCache(object):
+    """
+    This caches schema requests in redis and uses a method to
+    serve stale values while the cache is being populated or
+    updated to handle the thundering herd problem.
+    """
+
+    timeout = settings.SCHEMAS_REFRESH_SCHEDULE * 60
+    # keeping the stale cached items for 10 minutes longer
+    # than its timeout to make sure repopulation can work
+    stale_cache_timeout = 60 * 10
+
+    def __init__(self, data_source):
+        self.data_source = data_source
+        self.client = redis_connection
+        self.cache_key = "data_source:schema:cache:{}".format(self.data_source.id)
+        self.lock_key = "{}:lock".format(self.cache_key)
+        self.fresh_key = "{}:fresh".format(self.cache_key)
+
+    def load(self):
+        """
+        When called will fetch all table and column metadata from
+        the database and serialize it with the TableMetadataSerializer.
+        """
+        # due to the unfortunate import time side effects of
+        # Redash's package layout this needs to be done inline
+        from redash.serializers import TableMetadataSerializer
+
+        schema = []
+        tables = (
+            TableMetadata.query.filter(
+                TableMetadata.data_source_id == 1, TableMetadata.exists.is_(True),
```

Comment:
`TableMetadata.data_source_id == 1`? Did you forget to change something back after testing?

Reply:
OMG yeah, that's fixed in https://github.com/mozilla/redash/pull/995/commits/42852f74c4c564e4d8a13cd43e111ed4529bee2b again. That was just me copying things from the terminal, sigh.